subsys/bluetooth/mesh/access.c

```diff
@@ -2443,187 +2608,236 @@ static int read_comp_cb(const char *key, size_t len, settings_read_cb read_cb,
-	if (err) {
-		LOG_ERR("Failed reading models metadata: %d", err);
-		return err;
+	if (page >= 128) {
+		err = stored_page_read(buf, type, page, offset, allow_partial_elems);
+		if (err != -ENOENT) {
+			return err;
+		}
 	}
+#endif
+	return current_page_contents(buf, type, page % 128, offset, allow_partial_elems);
+}
 
-	/* First two bytes are total length */
-	offset += 2;
+size_t bt_mesh_comp_page_size(uint8_t page)
+{
+	return page_size_get(PAGE_TYPE_COMP, page);
+}
 
-	net_buf_simple_add_mem(buf, &stored_buf.data, MIN(net_buf_simple_tailroom(buf), 2));
+size_t bt_mesh_metadata_page_size(uint8_t page)
+{
+	return page_size_get(PAGE_TYPE_METADATA, page);
+}
 
-	if (offset >= stored_buf.len) {
-		return 0;
-	}
+bool bt_mesh_comp_128_changed(void)
+{
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	return stored_page_size_get(PAGE_TYPE_COMP, 128) != 0;
+#else
+	return false;
+#endif
+}
 
-	net_buf_simple_add_mem(buf, &stored_buf.data[offset],
-			       MIN(net_buf_simple_tailroom(buf), stored_buf.len - offset));
+int bt_mesh_comp_data_get_elems(struct net_buf_simple *buf, uint8_t page)
+{
+	return get_page_contents(buf, PAGE_TYPE_COMP, page, 0, false);
+}
 
-	LOG_DBG("metadata read %d", buf->len);
+int bt_mesh_comp_data_get_page(struct net_buf_simple *buf, uint8_t page, size_t offset)
+{
+#ifdef CONFIG_BT_MESH_LARGE_COMP_DATA_SRV
+	return get_page_contents(buf, PAGE_TYPE_COMP, page, offset, true);
+#else
+	return -EINVAL;
+#endif
+}
 
-	if (buf->len == original_len) {
-		return -ENOENT;
-	}
+int bt_mesh_metadata_get_page(struct net_buf_simple *buf, uint8_t page, size_t offset)
+{
+#ifdef CONFIG_BT_MESH_LARGE_COMP_DATA_SRV
+	return get_page_contents(buf, PAGE_TYPE_METADATA, page, offset, true);
+#else
+	return -EINVAL;
+#endif
+}
 
-	return 0;
+int bt_mesh_comp_data_set(uint8_t page, const void *data, uint16_t len)
+{
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	return stored_page_write(PAGE_TYPE_COMP, page, data, len);
+#else
+	return -ENOTSUP;
+#endif
 }
+
+int bt_mesh_metadata_set(uint8_t page, const void *data, uint16_t len)
+{
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	return stored_page_write(PAGE_TYPE_METADATA, page, data, len);
+#else
+	return -ENOTSUP;
 #endif
+}
 
-static void models_metadata_clear(void)
+int bt_mesh_model_data_store(const struct bt_mesh_model *mod, bool vnd, const char *name,
+			     const void *data, size_t data_len)
 {
+	char path[30];
 	int err;
 
-	err = settings_delete("bt/mesh/metadata");
-	if (err) {
-		LOG_ERR("Failed to clear models metadata: %d", err);
+	encode_mod_path(mod, vnd, "data", path, sizeof(path));
+	if (name) {
+		strcat(path, "/");
+		strncat(path, name, SETTINGS_MAX_DIR_DEPTH);
+	}
+
+	if (data_len) {
+		err = settings_save_one(path, data, data_len);
 	} else {
-		LOG_DBG("Cleared models metadata");
+		err = settings_delete(path);
 	}
 
-	atomic_clear_bit(bt_mesh.flags, BT_MESH_METADATA_DIRTY);
+	if (err) {
+		LOG_ERR("Failed to store %s value", path);
+	} else {
+		LOG_DBG("Stored %s value", path);
+	}
```

Comment:
This has nothing to do with my PR, it just looks like it was changed because I moved things around but if you look at the current main you will find this entire function like it is here (https://github.com/zephyrproject-rtos/zephyr/blob/0a7908588c755d3b8dfa3cc5c967d8a085762500/subsys/bluetooth/mesh/access.c#L2478-L2503). I would prefer to not fix all random bugs in the access module just becaused I change some parts.